File: src/003-largest-prime-factor/python/prime.py
import math

class Prime(object):
   def __init__(self):
      self.primes = [2]
      self.prime  = 2

   def check(self, prime):
      root = math.sqrt(prime)
      for d in self.primes:
         if d > root:
            return True
         if (prime % d) == 0:
            return False
      return True

   def next(self):
      if self.prime == 2:
         self.prime = 3
         return 2
      self.primes.append(self.prime)

      # find next prime
      self.prime += 2
      while not self.check(self.prime):
         self.prime += 2

      return self.primes[-1]

   def nth(self, n):
      while len(self.primes) < n:
         self.next()
      return self.primes[n-1]

# generator function for primes
def iter():
   p = Prime()
   while True:
      yield p.next()

File: src/ruby/007/src/euler/prime.rb
module Euler
  class Prime
    include Enumerable

    def initialize
      @primes = [2, 3]
    end

    def nth(p)
      self.next while @primes.length <= p

      @primes[p]
    end

    def next
      cur = @primes.last + 2
      cur += 2 until check_prime?(cur)
      @primes.push(cur)
      cur
    end

    def each
      @primes.each do |p|
        yield p
      end

      loop do
        yield self.next
      end
    end

    def check_prime?(n)
      max = Math.sqrt(n).floor

      @primes.each do |p|
        return true if p > max
        return false if (n % p) == 0
      end

      true
    end

    def check?(n)
      max = Math.sqrt(n).to_i

      self.next while @primes.last <= max

      check_prime? n
    end

    def index(n)
      self.next while n > @primes.last

      i = @primes.bsearch_index { |p, _| p >= n }
      i if i.nil? || @primes[i] == n
    end
  end
end
